cores/rp2040/Tone.cpp

```diff
@@ -48,6 +48,11 @@ void tone(uint8_t pin, unsigned int frequency, unsigned long duration) {
         return;
     }
 
+    auto entry = _toneMap.find(pin);
```

Comment:
Thanks for the PR.  I don’t know how I missed it.

Unfortunately, though, this section needs to be kept in a mutex protected block because the std::map is not thread safe.  A second Subblock could be done with the CoreMutex here, just using brackets to set the mutex lifetime to only be the find and end check.

Reply:
It's not elegant, but something like this is what I was suggesting:

````
            bool needStop;
            {
                CoreMutex m(&_toneMutex);
                auto entry = _toneMap.find(pin);
                needStop = entry != _toneMap.end();
            }
            if (needStop) {
                noTone(pin);
            }
````

Reply:
that works but not really explicit, how about we add explicit enter()/exit() for mutex. There maybe more of this situation later on. Though it is totally up to you.

Reply:
FWIW I think the CoreMutex class is actually more explicit/simpler than the enter/exit.  It's a wrapper that, for the object lifetime, grabs the mutex and releases it on any exit of the block.  It's got the anti-deadlock stuff from `pico-sdk`, which is what the `operator bool` is doing, although I'm thinking having it immediately `panic()` on deadlock may instead be the right thing to do...

Reply:
I agreed, I mean we should add .enter() and .exit() to the CoreMutex. For this case, we can actively release the mutex before  calling the noTone() and re-claim it afterward. That would make it easier to understand than replying on the destructor()

https://github.com/earlephilhower/arduino-pico/blob/master/cores/rp2040/CoreMutex.h#L50